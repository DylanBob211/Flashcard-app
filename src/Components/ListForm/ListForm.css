.listForm_container {
  display: flex;
  justify-content: center;
  align-items: center;
  border: 2px #c9c9c920 solid;
  border-radius: 10px;
  background: var(--bg-color);
  box-shadow: -9px 17px 21px -16px rgba(0,0,0,0.49);
}

.listForm_input-container {
  display: grid;
  height: 20%;
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: repeat(4, 1fr);
  align-items: center;
  justify-items: center;
}

.listForm_btn {
  fill: var(--secondary-color);
  color: var(--secondary-color);
  background: none;
  border: none;
  cursor: pointer;
  will-change: transform, opacity;
}

.listForm_btn--lg {
  height: 5rem;
  width: 5rem;
}

.listForm_btn--sm {
  height: 2rem;
  width: 2rem;
}

.listForm_btn--remove {
  grid-row-start: 3;
  grid-row-end: 5;
  grid-column-start: 1;
  grid-column-end: 3;
}

.listForm_btn--submit {
  grid-row-start: 3;
  grid-row-end: 5;
  grid-column-start: 3;
  grid-column-end: 5;
  font-size: 1.5rem;
}

.listForm_btn:hover{
  transform: scale(1.2);  
  fill: var(--third-shade-secondary)
}

.listForm_input {
  max-width: 70%;
  grid-column-start: 1;
  grid-column-end: 5;
  grid-row-start: 1;
  grid-row-end: 3;
  border: 3px #c9c9c990 solid;
  border-radius: 5px;
  height: 1.6rem;
  font-size: 1.5rem;
  will-change: transform, opacity;
}



/* Extra small devices (phones, 600px and down) */
@media only screen and (max-width: 600px) {

} 
/* Small devices (portrait tablets and large phones, 600px and up) */
@media only screen and (min-width: 600px) {
    
} 

/* Medium devices (landscape tablets, 768px and up) */
@media only screen and (min-width: 768px) {
   
} 

/* Large devices (laptops/desktops, 992px and up) */
@media only screen and (min-width: 992px) {
    
} 

/* Extra large devices (large laptops and desktops, 1200px and up) */
@media only screen and (min-width: 1200px) {
    
}