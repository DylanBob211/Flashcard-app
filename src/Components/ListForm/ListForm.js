import React, {useState} from 'react'
import './ListForm.css'

const ListForm = ({addNewList}) => {
    const [newList, setNewList] = useState({
        parole: [],
        name : ''
    })

    const createNewList = (e) => {
        e.preventDefault();
        addNewList(newList)
        setNewList({name : ''})
    }

    const handleChange = e => {
        setNewList({parole : [], name : e.target.value})
        
    }
    
    return (
        <>
            <form className="listForm_container" onSubmit={e => createNewList(e)}>
                <h1 className="listForm_title">Flashcard App</h1>
                <div className="listForm_input-container">
                    <input className="listForm_input" type="text" onChange={e => handleChange(e)} value={newList.name}></input>
                    <input className="listForm_btn" type="submit" value='+ Add Topic'></input>
                </div>
            </form>
        </>
    )
}

export default ListForm

