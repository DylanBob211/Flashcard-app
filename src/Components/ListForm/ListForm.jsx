import React, { useRef } from 'react';
import uuidv4 from 'uuidv4';
import PropTypes from 'prop-types';
import { useTransition, animated } from 'react-spring';
import './ListForm.css';
import PlusIcon from '../Icons/PlusIcon';
import RemoveIcon from '../Icons/RemoveIcon';

const ListForm = ({ addNewList, handleError }) => {
  const [newList, setNewList] = React.useState({
    words: [
      { url: [], word: '' },
    ],
    name: '',
    id: '',
  });
  const listNameInput = useRef(null);

  const [isInputting, setIsInputting] = React.useState(false);

  const openAddListTextInput = async () => {
    await setIsInputting(true);
    listNameInput.current.focus();
  };
  const closeAddListTextInput = () => {
    handleError('');
    setNewList(state => ({ ...state, name: '', id: '' }));
    setIsInputting(false);
  };
  const createNewList = (e) => {
    e.preventDefault();
    if (!newList.name) {
      handleError('Input a name for the list');
    } else {
      addNewList(newList);
      setNewList(state => ({ ...state, name: '', id: '' }));
      closeAddListTextInput();
    }
  };
  const handleChange = (e) => {
    e.persist();
    handleError('');
    setNewList({ id: uuidv4(), words: [], name: e.target.value });
  };

  const AnimatedPlusIcon = animated(PlusIcon);

  const transitions = useTransition(!isInputting, null, {
    from: { opacity: 0, transform: 'translate3d(100%,0,0)' },
    enter: { opacity: 1, transform: 'translate3d(0%,0,0)' },
    leave: { opacity: 0, transform: 'translate3d(-50%,0,0)' },
  });

  return (
    <div className="listForm_container">
      {
      transitions.map(({ item, key, props }) => (item ? (
        <AnimatedPlusIcon
          key={`${key} btn`}
          style={props}
          onClick={openAddListTextInput}
          data-test="addNewListButton"
          className="listForm_btn listForm_btn--lg"
        />
      ) : (
        <animated.form
          key={`${key} form`}
          style={props}
          className="listForm_input-container"
          data-test="addNewListForm"
          onSubmit={e => createNewList(e)}
        >
          <input
            ref={listNameInput}
            data-test="newListTextInput"
            className="listForm_input"
            type="text"
            onChange={e => handleChange(e)}
            value={newList.name}
          />
          <RemoveIcon
            className="listForm_btn listForm_btn--remove listForm_btn--sm"
            onClick={closeAddListTextInput}
            data-test="closeNewListInputButton"
          />
          <input
            data-test="submitNewListButton"
            className="listForm_btn listForm_btn--submit listForm_btn--sm"
            type="submit"
            value=">>"
          />
        </animated.form>
      )))
      }
    </div>
  );
};

ListForm.propTypes = {
  addNewList: PropTypes.func.isRequired,
  handleError: PropTypes.func.isRequired,
};

export default ListForm;
