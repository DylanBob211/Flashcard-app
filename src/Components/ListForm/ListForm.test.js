import React from 'react';
import { mount } from 'enzyme';
import ListForm from './ListForm';

describe('ListForm Component', () => {
  let component;
  const addNewListMock = jest.fn();
  const handleErrorMock = jest.fn();
  // const setState = jest.fn();
  // const useState = jest.spyOn(React, 'useState');
  // useState.mockImplementation(init => [init, setState]);
  beforeEach(() => {
    component = mount(<ListForm addNewList={addNewListMock} handleError={handleErrorMock} />);
    component
      .find('svg[data-test="addNewListButton"]')
      .simulate('click');
  });

  afterEach(() => {
    jest.clearAllMocks();
    component.unmount();
  });

  // while not possible to properly test useState
  it.skip('calls setIsInputting when the plusIcon is pressed', () => {
    component
      .find('svg[data-test="addNewListButton"]')
      .simulate('click');
    expect(setState).toHaveBeenCalledWith(true);
  });

  it('focuses the inputForm when plusButton is pressed', () => {
    const input = component.find('[data-test="newListTextInput"]');
    expect(input.is(':focus')).toBe(true);
  });

  it('calls handleError if no input is submitted', () => {
    component
      .find('[data-test="submitNewListButton"]')
      .simulate('submit');
    expect(addNewListMock).not.toHaveBeenCalled();
    expect(handleErrorMock).toHaveBeenCalledWith('Input a name for the list');
  });

  it('calls addNewList only if some input is submitted', () => {
    component
      .find('[data-test="newListTextInput"]')
      .simulate('change', { target: { value: 'NewList' } });
    component
      .find('[data-test="submitNewListButton"]')
      .simulate('submit');

    expect(addNewListMock).toHaveBeenCalled();
  });

  it('calls handleError with empty string when typing into the input component', () => {
    component
      .find('[data-test="newListTextInput"]')
      .simulate('change', { target: { value: 'NewList' } });

    expect(handleErrorMock).toHaveBeenCalledWith('');
  });
});
