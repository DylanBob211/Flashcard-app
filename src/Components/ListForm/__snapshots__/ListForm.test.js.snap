// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListForm Component Matches to snapshot 1`] = `
<ListForm
  addNewList={[MockFunction]}
  handleError={[MockFunction]}
>
  <div
    className="listForm_container"
  >
    <ForwardRef
      className="listForm_btn listForm_btn--lg"
      data-test="addNewListButton"
      key="0 btn"
      onClick={[Function]}
      style={
        Object {
          "opacity": AnimatedValue {
            "animatedStyles": Set {},
            "children": Array [
              AnimatedStyle {
                "attach": [Function],
                "children": Array [
                  AnimatedProps {
                    "attach": [Function],
                    "children": Array [],
                    "detach": [Function],
                    "payload": Object {
                      "className": "listForm_btn listForm_btn--lg",
                      "data-test": "addNewListButton",
                      "onClick": [Function],
                      "style": [Circular],
                    },
                    "update": [Function],
                  },
                ],
                "detach": [Function],
                "payload": [Circular],
              },
            ],
            "done": true,
            "lastPosition": 0,
            "lastTime": undefined,
            "lastVelocity": undefined,
            "payload": undefined,
            "setValue": [Function],
            "startPosition": 0,
            "startTime": 1566892965061,
            "value": 0,
          },
          "position": AnimatedValue {
            "animatedStyles": Set {},
            "children": Array [
              AnimatedStyle {
                "attach": [Function],
                "children": Array [
                  AnimatedProps {
                    "attach": [Function],
                    "children": Array [],
                    "detach": [Function],
                    "payload": Object {
                      "className": "listForm_btn listForm_btn--lg",
                      "data-test": "addNewListButton",
                      "onClick": [Function],
                      "style": [Circular],
                    },
                    "update": [Function],
                  },
                ],
                "detach": [Function],
                "payload": [Circular],
              },
            ],
            "done": true,
            "lastPosition": "absolute",
            "lastTime": undefined,
            "lastVelocity": undefined,
            "payload": undefined,
            "setValue": [Function],
            "startPosition": "absolute",
            "startTime": 1566892965058,
            "value": "absolute",
          },
          "transform": AnimatedInterpolation {
            "attach": [Function],
            "calc": [Function],
            "children": Array [
              AnimatedStyle {
                "attach": [Function],
                "children": Array [
                  AnimatedProps {
                    "attach": [Function],
                    "children": Array [],
                    "detach": [Function],
                    "payload": Object {
                      "className": "listForm_btn listForm_btn--lg",
                      "data-test": "addNewListButton",
                      "onClick": [Function],
                      "style": [Circular],
                    },
                    "update": [Function],
                  },
                ],
                "detach": [Function],
                "payload": [Circular],
              },
            ],
            "detach": [Function],
            "payload": Array [
              AnimatedValue {
                "animatedStyles": Set {},
                "children": Array [
                  [Circular],
                ],
                "done": false,
                "lastPosition": 0,
                "lastTime": undefined,
                "lastVelocity": undefined,
                "payload": undefined,
                "setValue": [Function],
                "startPosition": 0,
                "startTime": 1566892965097,
                "value": 0,
              },
            ],
          },
        }
      }
    >
      <PlusIcon
        className="listForm_btn listForm_btn--lg"
        data-test="addNewListButton"
        onClick={[Function]}
        style={
          Object {
            "opacity": 0,
            "position": "absolute",
            "transform": "translate3d(100%,0,0)",
          }
        }
      >
        <svg
          className="icon listForm_btn listForm_btn--lg"
          data-test="addNewListButton"
          height="24"
          onClick={[Function]}
          style={
            Object {
              "opacity": 0,
              "position": "absolute",
              "transform": "translate3d(100%,0,0)",
            }
          }
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12zm6 13h-5v5h-2v-5h-5v-2h5v-5h2v5h5v2z"
          />
        </svg>
      </PlusIcon>
    </ForwardRef>
    <ForwardRef
      className="listForm_input-container"
      data-test="addNewListForm"
      key="1 form"
      onSubmit={[Function]}
      style={
        Object {
          "opacity": AnimatedValue {
            "animatedStyles": Set {},
            "children": Array [
              AnimatedStyle {
                "attach": [Function],
                "children": Array [
                  AnimatedProps {
                    "attach": [Function],
                    "children": Array [],
                    "detach": [Function],
                    "payload": Object {
                      "children": Array [
                        <input
                          className="listForm_input"
                          data-test="newListTextInput"
                          onChange={[Function]}
                          type="text"
                          value=""
                        />,
                        <RemoveIcon
                          className="listForm_btn listForm_btn--remove listForm_btn--sm"
                          data-test="closeNewListInputButton"
                          onClick={[Function]}
                        />,
                        <input
                          className="listForm_btn listForm_btn--submit listForm_btn--sm"
                          data-test="submitNewListButton"
                          type="submit"
                          value=">>"
                        />,
                      ],
                      "className": "listForm_input-container",
                      "data-test": "addNewListForm",
                      "onSubmit": [Function],
                      "style": [Circular],
                    },
                    "update": [Function],
                  },
                ],
                "detach": [Function],
                "payload": [Circular],
              },
            ],
            "done": false,
            "lastPosition": 0,
            "lastTime": undefined,
            "lastVelocity": undefined,
            "payload": undefined,
            "setValue": [Function],
            "startPosition": 0,
            "startTime": 1566892965097,
            "value": 0,
          },
          "position": AnimatedValue {
            "animatedStyles": Set {},
            "children": Array [
              AnimatedStyle {
                "attach": [Function],
                "children": Array [
                  AnimatedProps {
                    "attach": [Function],
                    "children": Array [],
                    "detach": [Function],
                    "payload": Object {
                      "children": Array [
                        <input
                          className="listForm_input"
                          data-test="newListTextInput"
                          onChange={[Function]}
                          type="text"
                          value=""
                        />,
                        <RemoveIcon
                          className="listForm_btn listForm_btn--remove listForm_btn--sm"
                          data-test="closeNewListInputButton"
                          onClick={[Function]}
                        />,
                        <input
                          className="listForm_btn listForm_btn--submit listForm_btn--sm"
                          data-test="submitNewListButton"
                          type="submit"
                          value=">>"
                        />,
                      ],
                      "className": "listForm_input-container",
                      "data-test": "addNewListForm",
                      "onSubmit": [Function],
                      "style": [Circular],
                    },
                    "update": [Function],
                  },
                ],
                "detach": [Function],
                "payload": [Circular],
              },
            ],
            "done": true,
            "lastPosition": "absolute",
            "lastTime": undefined,
            "lastVelocity": undefined,
            "payload": undefined,
            "setValue": [Function],
            "startPosition": "absolute",
            "startTime": 1566892965097,
            "value": "absolute",
          },
          "transform": AnimatedInterpolation {
            "attach": [Function],
            "calc": [Function],
            "children": Array [
              AnimatedStyle {
                "attach": [Function],
                "children": Array [
                  AnimatedProps {
                    "attach": [Function],
                    "children": Array [],
                    "detach": [Function],
                    "payload": Object {
                      "children": Array [
                        <input
                          className="listForm_input"
                          data-test="newListTextInput"
                          onChange={[Function]}
                          type="text"
                          value=""
                        />,
                        <RemoveIcon
                          className="listForm_btn listForm_btn--remove listForm_btn--sm"
                          data-test="closeNewListInputButton"
                          onClick={[Function]}
                        />,
                        <input
                          className="listForm_btn listForm_btn--submit listForm_btn--sm"
                          data-test="submitNewListButton"
                          type="submit"
                          value=">>"
                        />,
                      ],
                      "className": "listForm_input-container",
                      "data-test": "addNewListForm",
                      "onSubmit": [Function],
                      "style": [Circular],
                    },
                    "update": [Function],
                  },
                ],
                "detach": [Function],
                "payload": [Circular],
              },
            ],
            "detach": [Function],
            "payload": Array [
              AnimatedValue {
                "animatedStyles": Set {},
                "children": Array [
                  [Circular],
                ],
                "done": false,
                "lastPosition": 0,
                "lastTime": undefined,
                "lastVelocity": undefined,
                "payload": undefined,
                "setValue": [Function],
                "startPosition": 0,
                "startTime": 1566892965097,
                "value": 0,
              },
            ],
          },
        }
      }
    >
      <form
        className="listForm_input-container"
        data-test="addNewListForm"
        onSubmit={[Function]}
        style={
          Object {
            "opacity": 0,
            "position": "absolute",
            "transform": "translate3d(100%,0,0)",
          }
        }
      >
        <input
          className="listForm_input"
          data-test="newListTextInput"
          onChange={[Function]}
          type="text"
          value=""
        />
        <RemoveIcon
          className="listForm_btn listForm_btn--remove listForm_btn--sm"
          data-test="closeNewListInputButton"
          onClick={[Function]}
        >
          <svg
            className="icon listForm_btn listForm_btn--remove listForm_btn--sm"
            data-test="closeNewListInputButton"
            height="24"
            onClick={[Function]}
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12zm6 13h-12v-2h12v2z"
            />
          </svg>
        </RemoveIcon>
        <input
          className="listForm_btn listForm_btn--submit listForm_btn--sm"
          data-test="submitNewListButton"
          type="submit"
          value=">>"
        />
      </form>
    </ForwardRef>
  </div>
</ListForm>
`;
